### ISSUE TITLE
Issues with up & download

### ISSUE THREAD

**Issue by thomas-br** [brightio/penelope#28]
I am facing issues using the up- & download feature.

For uploads I constantly receive the following error:

```
Traceback (most recent call last):
  File "/usr/lib/python3.11/threading.py", line 1045, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.11/threading.py", line 982, in run
    self._target(*self._args, **self._kwargs)
  File "/mnt/hgfs/tools/penelope.py", line 365, in cmdloop
    stop = self.onecmd(line)
           ^^^^^^^^^^^^^^^^^
  File "/mnt/hgfs/tools/penelope.py", line 376, in onecmd
    return func(arg)
           ^^^^^^^^^
  File "/mnt/hgfs/tools/penelope.py", line 581, in newfunc
    return func(self, ID)
           ^^^^^^^^^^^^^^
  File "/mnt/hgfs/tools/penelope.py", line 852, in do_upload
    core.sessions[self.sid].upload(local_items, randomize_fname=True)
  File "/mnt/hgfs/tools/penelope.py", line 3168, in upload
    cwd_escaped = self.cwd.replace('\\', '\\\\')
                  ^^^^^^^^^^^^^^^^
AttributeError: 'bool' object has no attribute 'replace'

Penelope version: 0.12.2
Python version: 3.11.9 (main, Apr 10 2024, 13:16:36) [GCC 13.2.0]
System: #1 SMP PREEMPT_DYNAMIC Kali 6.6.9-1kali1 (2024-01-08)
```

When using download, I see the following error in the windows shell:

```
iex : At line:2 char:158
+ ... wHqEbPRJnrjJrrb.bat" "%TEMP%\SwHqEbPRJnrjJrrb.bat" >NUL 2>&1&"%TEMP%\ ...
+                                                                 ~
The ampersand (&) character is not allowed. The & operator is reserved for future use; wrap an ampersand in double
quotation marks ("&") to pass it as part of a string.
At line:2 char:188
+ ... wHqEbPRJnrjJrrb.bat" >NUL 2>&1&"%TEMP%\SwHqEbPRJnrjJrrb.bat"&del "%TE ...
+                                                                 ~
The ampersand (&) character is not allowed. The & operator is reserved for future use; wrap an ampersand in double
quotation marks ("&") to pass it as part of a string.
At line:1 char:289
+ ... Encoding).GetString($bytes,0, $i);$sendback = (iex $data 2>&1 | Out-S ...
+                                                    ~~~~~~~~~~~~~~
    + CategoryInfo          : ParserError: (:) [Invoke-Expression], ParseException
    + FullyQualifiedErrorId : AmpersandNotAllowed,Microsoft.PowerShell.Commands.InvokeExpressionCommand
```

Any ideas?

**Comment by brightio** [comment]
This should be fixed now with the commit 12468b0. Can you please try now? Thanks for reporting!

**Comment by thomas-br** [comment]
@brightio works now thanks!

But another thing it noticed: On Windows the shell is not constantly returning the results but only after finishing the command. This can be demoed easily using `ping <ip> -n 10`, which only shows the results after all 10 pings completed. This is quite inconvenient when e.g. running winPEAS.exe which takes a while and you cannot use the time to inspect the output returned so far. Shall I open a separate issue for this?

**Comment by brightio** [comment]
Can you open a different issue for this please? I will look at it asap

**Comment by funkyfliam** [comment]
I believe I'm still getting this error after updating to latest version.

```
PS C:\Windows\Temp\feel> 
[!] Session detached...

┍┽ penelope ┾┑ Session [2] > upload https://raw.githubusercontent.com/antonioCoco/ConPtyShell/master/Invoke-ConPtyShell.ps1
[+] --- ⇣  Downloading https://raw.githubusercontent.com/antonioCoco/ConPtyShell/master/Invoke-ConPtyShell.ps1

Oops...🐞
────────────────────────────────────────────────────────────────────────────────
Traceback (most recent call last):                                                                                                                                                           
  File "/usr/lib/python3.12/threading.py", line 1075, in _bootstrap_inner                                                                                                                    
    self.run()                                                                                                                                                                               
  File "/usr/lib/python3.12/threading.py", line 1012, in run                                                                                                                                 
    self._target(*self._args, **self._kwargs)                                                                                                                                                
  File "/home/kali/ck19apo102/penelope.py", line 380, in cmdloop                                                                                                                  
    stop = self.onecmd(line)                                                                                                                                                                 
           ^^^^^^^^^^^^^^^^^                                                                                                                                                                 
  File "/home/kali/ck19apo102/penelope.py", line 392, in onecmd                                                                                                                   
    return func(arg)                                                                                                                                                                         
           ^^^^^^^^^                                                                                                                                                                         
  File "/home/kali/ck19apo102/penelope.py", line 606, in newfunc                                                                                                                  
    return func(self, ID)                                                                                                                                                                    
           ^^^^^^^^^^^^^^                                                                                                                                                                    
  File "/home/kali/ck19apo102/penelope.py", line 882, in do_upload                                                                                                                
    core.sessions[self.sid].upload(local_items, randomize_fname=True)                                                                                                                        
  File "/home/kali/ck19apo102/penelope.py", line 3259, in upload                                                                                                                  
    cwd_escaped = self.cwd.replace('\\', '\\\\')                                                                                                                                             
                  ^^^^^^^^^^^^^^^^                                                                                                                                                           
AttributeError: 'bool' object has no attribute 'replace'                                                                                                                                     
────────────────────────────────────────────────────────────────────────────────                                                                                                             
Penelope version: 0.12.4                                                                                                                                                                     
Python version: 3.12.6 (main, Sep  7 2024, 14:20:15) [GCC 14.2.0]
System: #1 SMP PREEMPT_DYNAMIC Kali 6.10.9-1kali1 (2024-09-09)
```


--edit
Seems to occur when the target shell is running Powershell. CWD = False at time of error. Error below


Error - 
```
b't line:1 char:26\r\n+ set YCJmiWilhR=fcSKVhyJeg&set oUmCTGUMnP=JckBcyihjO\r\n+                          ~\r\nThe ampersand (&) character is not allowed. The & operator is reserved for future use; wrap an ampersand in double \r\nquotation marks ("&") to pass it as part of a string.\r\n    + CategoryInfo          : ParserError: (:) [], ParentContainsErrorRecordException\r\n    + FullyQualifiedErrorId : AmpersandNotAllowed\r\n \r\nPS C:\\Windows\\Temp\\feel> '
```

**Comment by brightio** [comment]
@funkyfliam I can't reproduce this error. Can you please tell me which powershell payload did you use to get the shell?
